source/adapters/level_zero/command_buffer.cpp

```diff
@@ -982,20 +1050,275 @@ UR_APIEXPORT ur_result_t UR_APICALL urCommandBufferEnqueueExp(
+  UR_ASSERT(Command->CommandBuffer->IsUpdatable,
+            UR_RESULT_ERROR_INVALID_OPERATION);
+  UR_ASSERT(Command->CommandBuffer->IsFinalized,
+            UR_RESULT_ERROR_INVALID_OPERATION);
+
+  auto CommandBuffer = Command->CommandBuffer;
+  uint32_t Dim = CommandDesc->newWorkDim;
+  const void *NextDesc = nullptr;
+
+  // We need the created descriptors to live till the point when
+  // zexCommandListUpdateMutableCommandsExp is called at the end of the
+  // function.
+  std::vector<std::unique_ptr<ZeStruct<ze_mutable_kernel_argument_exp_desc_t>>>
+      ArgDescs;
+  std::vector<std::unique_ptr<ZeStruct<ze_mutable_global_offset_exp_desc_t>>>
+      OffsetDescs;
+  std::vector<std::unique_ptr<ZeStruct<ze_mutable_group_size_exp_desc_t>>>
+      GroupSizeDescs;
+  std::vector<std::unique_ptr<ZeStruct<ze_mutable_group_count_exp_desc_t>>>
+      GroupCountDescs;
+
+  // Check if new global offset is provided.
+  size_t *NewGlobalWorkOffset = CommandDesc->pNewGlobalWorkOffset;
+  if (NewGlobalWorkOffset && Dim > 0) {
+    if (!CommandBuffer->Context->getPlatform()
+             ->ZeDriverGlobalOffsetExtensionFound) {
+      urPrint("No global offset extension found on this driver\n");
+      return UR_RESULT_ERROR_INVALID_VALUE;
+    }
+    auto MutableGroupOffestDesc =
+        std::make_unique<ZeStruct<ze_mutable_global_offset_exp_desc_t>>();
+    MutableGroupOffestDesc->commandId = Command->CommandId;
+    MutableGroupOffestDesc->pNext = NextDesc;
+    MutableGroupOffestDesc->offsetX = NewGlobalWorkOffset[0];
+    MutableGroupOffestDesc->offsetY = Dim >= 2 ? NewGlobalWorkOffset[1] : 0;
+    MutableGroupOffestDesc->offsetZ = Dim == 3 ? NewGlobalWorkOffset[2] : 0;
+    NextDesc = MutableGroupOffestDesc.get();
+    OffsetDescs.push_back(std::move(MutableGroupOffestDesc));
+  }
+
+  // Check if new group size is provided.
+  size_t *NewLocalWorkSize = CommandDesc->pNewLocalWorkSize;
+  if (NewLocalWorkSize && Dim > 0) {
+    auto MutableGroupSizeDesc =
+        std::make_unique<ZeStruct<ze_mutable_group_size_exp_desc_t>>();
+    MutableGroupSizeDesc->commandId = Command->CommandId;
+    MutableGroupSizeDesc->pNext = NextDesc;
+    MutableGroupSizeDesc->groupSizeX = NewLocalWorkSize[0];
+    MutableGroupSizeDesc->groupSizeY = Dim >= 2 ? NewLocalWorkSize[1] : 1;
+    MutableGroupSizeDesc->groupSizeZ = Dim == 3 ? NewLocalWorkSize[2] : 1;
+    NextDesc = MutableGroupSizeDesc.get();
+    GroupSizeDescs.push_back(std::move(MutableGroupSizeDesc));
+  }
+
+  // Check if new global size is provided and we need to update group count.
+  size_t *NewGlobalWorkSize = CommandDesc->pNewGlobalWorkSize;
+  if (NewGlobalWorkSize && Dim > 0) {
+    ze_group_count_t ZeThreadGroupDimensions{1, 1, 1};
+    uint32_t WG[3];
+    // If new global work size is provided but new local work size is not
+    // provided then we still need to update local work size based on size
+    // suggested by the driver for the kernel.
```

Comment:
In my mind the expected semantics is that if the global work size changes, and local work size is not provided then the local work size stays the same and it's an error if the global/local sizes are not compatible. This is based on similar OpenCL semantics in [cl_mutable_dispatch_config_khr](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_Ext.html#_mutable_command_types)

> local_work_size Points to an array of work_dim unsigned values that describe the number of work-items that make up a work-group that will execute the kernel. If local_work_size is NULL then the number of local work-items in the dispatch is not changed. See clEnqueueNDRangeKernel for valid usage.

I'm not sure if that's the best semantics though, in our [SYCL-Graph Update API](https://github.com/intel/llvm/pull/12486) we can update the range based on a `sycl::range` or a `sycl::nd_range`. Using a `sycl::nd_range` would mean that both global & local work sizes are set, while using a `sycl::range` implies the user doesn't care about the local size.

Reply:
To recap my own perspective on this from the call. A SYCL user can't go from a `sycl::range` to an `sycl::nd_range` during update an vice versa. So they will either care about the local size or not. The current semantics of picking the best local size possible when a user hasn't set the local size I think is desirable, because it corresponds to the usecase where a user indicates they don't care about local size so would want the implementation to pick for them.

I'll take the action to refine the UR API to remove the ambiguity about what happens if a user creates a kernel with a local size, and then only updates the global size. But for this purpose of this PR I think current the implementation is fine.

Reply:
Thank you for clarifying!